src/resolv/lookup/srv.rs

```
@@ -0,0 +1,215 @@
+    resolver: Resolver,
+
+    /// The current querry.
+    query: Query,
+
+    /// An optional search list iterator for searching a name.
+    search: Option<SearchIter>,
+}
+
+
+//--- Future
+
+impl Future for LookupSRV {
+    type Item = FoundSRVs;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let err = match self.query.poll() {
+            Ok(Async::NotReady) => return Ok(Async::NotReady),
+            Ok(Async::Ready(item)) => return Ok(Async::Ready(FoundSRVs::new(item)?)),
+            Err(err) => err
+        };
+        let name = match self.search {
+            None => return Err(err),
+            Some(ref mut search) => {
+                match search.next() {
+                    None => return Err(err),
+                    Some(name) => name,
+                }
+            }
+        };
+        self.query = self.resolver.clone()
+                         .query((name, Rtype::Srv, Class::In));
+        self.poll()
+    }
+}
+
+
+//------------ FoundSRVs ----------------------------------------------------
+
+/// The value returned by a successful SRV lookup.
+///
+/// TODO: doc
+///
+/// The `canonical_name()` method returns the canonical name of the host for
+/// which the addresses were found.
+#[derive(Clone, Debug)]
+pub struct FoundSRVs {
+    /// The canonical domain name for the host.
+    canonical: DNameBuf,
+
+    /// All the IP addresses we’ve got.
+    results: Vec<SRVItem>
+}
+
+#[derive(Clone, Debug)]
+pub struct SRVItem {
+    host: DNameBuf,
+    port: u16,
+    addrs: Option<Vec<IpAddr>>
+}
+
+impl FoundSRVs {
+    /// Creates a new value from the results of the A and AAAA queries.
+    ///
+    /// Either of the queries can have resulted in an error but not both.
+    fn new(result: MessageBuf) -> Result<Self> {
+        let name = result.canonical_name().unwrap();
+        let mut rrs = Vec::new();
+        Self::process_records(&mut rrs, &result, &name).ok();
```

You should probably not ignore any error returned by `process_records()` here but rather “question mark out.”

I am meaning to rework the error types, but for the moment `resolv::error::Error` will do. I think it is semantically okay to convert the `ParseError` into a `std::io::Error` (with the `Other` kind) and use the `resolv::error::Error::Io` variant. If you add an `impl From<ParseError> for Error` doing that, you can simply use the question mark here and in all the other places where you currently call `ok()`.

---

Done.